### ISSUE TITLE
PBEM: Can no longer send messages using a Hotmail account

### ISSUE THREAD

**Issue by jarollins84** [triplea-game/triplea#2595]
### My Operating System:
Windows 10

### TripleA version:
1.9.0.0.7378

### Map:
None

### Can you describe how to trigger the error?
When I hit test email button

### Do you have the exact error text?
I can't get it all in or copy it but it starts like this unable to send email: 554 5.2.0 STOREDRV.Submission.Exception:sendAasDeniedExceptionSendAsDenied; Failed to processs message due to a permanent exception with message cannon submit message

### Instead of this error, what should have happened?
it should be sending me an test email

**Comment by ssoloff** [comment]
This is possibly due to a large save game being attached to the message (see [here](https://social.technet.microsoft.com/Forums/ie/en-US/a46f4323-f2f1-4b7b-96f0-9fbbea5d031e/554-520?forum=exchangesvrsecuremessaginglegacy)).

@jarollins84 Can you please zip and attach the save game you're trying to send to this issue?  Also, which email provider are you using?

**Comment by jarollins84** [comment]
Steven,


I wasn't even in a game yet I was just trying to test the email through the PBEM tab now the message is this.   Unable to send email: 451 4.7.0 Temporary server error. Please try again later.  PRX4 [MWHPR0201CA0052.namprd02.prod.outlook.com].  I have reinstalled the program and restarted computer, and am not very tech savy on how I am going to fix this if there isn't a real easy fix.


Alan


________________________________

**Comment by ssoloff** [comment]
@jarollins84 Ah, sorry, I read your description too quickly. :smile:

The latest error you posted indicates it may be a temporary problem with `outlook.com`.  I don't use `outlook.com`, but I'll create a throwaway account and see if I can reproduce the problem.

**Comment by jarollins84** [comment]
It went back to the old long error and i'm using a Hotmail email.


________________________________

**Comment by jarollins84** [comment]
Also trying to use the standard global 1940 second edition map


________________________________

**Comment by ssoloff** [comment]
Sweet!  While trying to send a test email using a Hotmail account on my Linux box, the JVM core dumped and my entire GNOME session was destroyed. 😞 

Switching to Windows, I can confirm that I'm getting a similar message to what you're reporting.

I found [this post](https://answers.microsoft.com/en-us/msoffice/forum/mso_win10/i-cannot-send-out-email-anymore-from-outlook/eb21d415-883f-4ad1-b69d-5f05007064b3?auth=1) which seems to indicate other Outlook.com users have started seeing this problem in recent months.  It looks like they hardened their security and will reject emails with an unverified From address:

> This applies to all email clients using smtp-mail.outlook.com to send email.  Sometime near the beginning of July, there was a change that disallowed you using a return email address that is not in your list of verified email aliases or connected accounts.
In other words, you can't change the "From" field in emails you send to an anonymous email address. 

I'll investigate later to see if there's a way to get this working again.  It sounds like it may be broken for all Outlook.com/Hotmail PBEM users.

**Comment by ssoloff** [comment]
>  While trying to send a test email using a Hotmail account on my Linux box, the JVM core dumped and my entire GNOME session was destroyed.

For my reference... this appears to be due to the fact that we're creating an excessively-wide dialog from the error message sent by Outlook.com.  It appears to be about five screens wide.  There may be a limitation in GNOME that's causing the core dump.  We may have to truncate and/or wrap the text to ensure the dialog width is reasonable.

**Comment by jarollins84** [comment]
Thank you I look forward to hearing from you and am about to start a gmail account lol.



________________________________

**Comment by ssoloff** [comment]
@DanVanAtta @RoiEXLab @ron-murhammer It appears the claim made at the above link regarding Hotmail rejecting messages with an unverified `From` field is correct.  I confirmed that I can get Hotmail PBEM working again by specifying the user's account name in the `From` field of the message:

```diff
diff --git a/src/main/java/games/strategy/engine/pbem/GenericEmailSender.java b/src/main/java/games/strategy/engine/pbem/GenericEmailSender.java
index 054fc2784..d44da5e72 100644
--- a/src/main/java/games/strategy/engine/pbem/GenericEmailSender.java
+++ b/src/main/java/games/strategy/engine/pbem/GenericEmailSender.java
@@ -142,7 +142,7 @@ public class GenericEmailSender implements IEmailSender {
     props.put("mail.smtp.connectiontimeout", m_timeout);
     props.put("mail.smtp.timeout", m_timeout);
     final String to = m_toAddress;
-    final String from = "noreply@triplea-game.org";
+    final String from = getUserName();
     // todo get the turn and player number from the game data
     try {
       final Session session = Session.getInstance(props, null);
```

The only issue I see here is from a privacy perspective in that the sending user's email is now being disclosed in the message sent by TripleA whereas it was not disclosed previously.  However, I don't think this is a legitimate concern because the other player(s) would have to know the sender's email address somehow in order to continue the game?  So one can't really remain anonymous in a PBEM game, correct?

**Comment by jarollins84** [comment]
Yeah I’m a pbem I would assume the players know each other.  Now what exactly did u do
Or do I need to do to fix it as I don’t understand what u have put in.  I’m not tech savvy just like using it and having others fix haha


Sent from my iPhone

On Nov 11, 2017, at 5:18 PM, Steven Soloff <notifications@github.com<mailto:notifications@github.com>> wrote:


@DanVanAtta<https://github.com/danvanatta> @RoiEXLab<https://github.com/roiexlab> @ron-murhammer<https://github.com/ron-murhammer> It appears the claim made at the above link regarding Hotmail rejecting messages with an unverified From field is correct. I confirmed that I can get Hotmail PBEM working again by specifying the user's account name in the From field of the message:

diff --git a/src/main/java/games/strategy/engine/pbem/GenericEmailSender.java b/src/main/java/games/strategy/engine/pbem/GenericEmailSender.java
index 054fc2784..d44da5e72 100644
--- a/src/main/java/games/strategy/engine/pbem/GenericEmailSender.java
+++ b/src/main/java/games/strategy/engine/pbem/GenericEmailSender.java
@@ -142,7 +142,7 @@ public class GenericEmailSender implements IEmailSender {
     props.put("mail.smtp.connectiontimeout", m_timeout);
     props.put("mail.smtp.timeout", m_timeout);
     final String to = m_toAddress;
-    final String from = "noreply@triplea-game.org<mailto:noreply@triplea-game.org>";
+    final String from = getUserName();
     // todo get the turn and player number from the game data
     try {
       final Session session = Session.getInstance(props, null);

The only issue I see here is from a privacy perspective in that the sending user's email is now being disclosed in the message sent by TripleA whereas it was not disclosed previously. However, I don't think this is a legitimate concern because the other player(s) would have to know the sender's email address somehow in order to continue the game? So one can't really remain anonymous in a PBEM game, correct?

**Comment by ssoloff** [comment]
@jarollins84 Unfortunately, there's nothing you can do on your end because it's going to require a code change to fix this issue.  Hopefully, we can get a fixed release out soon.  In the interim, your best bet is to use a Gmail account for PBEM.  Sorry for any inconvenience.

**Comment by jarollins84** [comment]
So tried it with gmail accound ja.rollins840815@gmail.com and it gave me another error.  I don't know what to do with this one how do I log in on my web browser.

[cid:27820dda-57d9-4221-86c5-08b8d165c422]


________________________________

**Comment by ssoloff** [comment]
Hmm, not really sure what that error code signifies.  Can you provide a screenshot of the error dialog TripleA pops up?

Are you able to login at https://gmail.com?  If so, do you know if you have two-factor authentication enabled?  If you are using two-factor authentication, you will need to create and use an "application password" for TripleA; see [here](https://support.google.com/accounts/answer/185833?hl=en) for more information.

**Comment by jarollins84** [comment]
Did you see what I sent?  if so that's all that the message said.  I also don't have two-factor authntiactiaon on should I turn it on.


________________________________

**Comment by ssoloff** [comment]
> [cid:27820dda-57d9-4221-86c5-08b8d165c422]

Just to be clear, you're saying TripleA displayed a dialog box with only the above text in the window?  What was the title of the dialog?  Normally, TripleA will prefix the error message received from the mail provider with something like "Unable to send email:".  That's why I'm confused when you said, "that's all the message said."

I just confirmed that I can PBEM using my Gmail account, so it appears to be working (i.e. it's not broken like Hotmail is).

> I also don't have two-factor authntiactiaon on should I turn it on.

That's not necessary for PBEM.  I was just curious as I know you have to use an application password for PBEM when you do have it enabled.  I don't remember if you have to use an application password even when two-factor authentication is disabled.

**Comment by jarollins84** [comment]
I sent the print screen let me know if you got it.  Also I couldn't find out how to give an application password when I went to the link it said my account didn't require that.


________________________________

**Comment by ssoloff** [comment]
> I sent the print screen let me know if you got it.

Where did you send it?  Did you attach it to an email you used to reply to this thread?  I don't know if GitHub strips out attachments when it converts your email reply into an issue comment.  If you use the [GitHub web interface](https://github.com/triplea-game/triplea/issues/2595), you can just drag-and-drop the image into your comment.

> I couldn't find out how to give an application password when I went to the link it said my account didn't require that.

Yeah, so it sounds like Google disables that feature if you don't have two-factor authentication enabled, so forget I mentioned it. :smile:

**Comment by jarollins84** [comment]
![screen shot](https://user-images.githubusercontent.com/33582631/32694791-86815814-c717-11e7-9268-54776bf08eae.png)

**Comment by jarollins84** [comment]
I will try to do two verification on it.

**Comment by jarollins84** [comment]
It looks like the two part verification and app password worked

**Comment by ssoloff** [comment]
Thank you.  The link in that message indicates Gmail is not allowing you to authenticate from TripleA.  From the list of things it says to try, this one sounds the most promising:

> Allow less secure apps: If you don't use 2-Step Verification, you might need to [allow less secure apps to access your account](https://support.google.com/accounts/answer/6010255).

Or you can just enable two-factor authentication and use an app password.  **EDIT:** Which apparently you've already done. :smile:

**Comment by jarollins84** [comment]
Thank you for your help I really appreciate it

**Comment by ssoloff** [comment]
@DanVanAtta @RoiEXLab @ron-murhammer 

> The only issue I see here is from a privacy perspective in that the sending user's email is now being disclosed in the message sent by TripleA whereas it was not disclosed previously.

I didn't notice this before, but even though we're currently using `noreply@triplea-game.org` in the `From` field, Gmail seems to always replace it with the addy of the authenticated user.  Thus, the proposed change above simply brings Hotmail (and the generic SMTP provider) into parity with Gmail.

**Comment by ssoloff** [comment]
@jarollins84 This should be fixed in 1.9.0.0.7514 or later.  Please reopen if you still have problems with PBEM using your Hotmail account.
